Source/Container/HCList.c

```diff
@@ -359,20 +359,47 @@ HCBoolean HCListIterationHasEnded(HCListIterator* iterator) {
 //----------------------------------------------------------------------------------------------------------------------------------
 // MARK: - Iteration Convenience Operations
 //----------------------------------------------------------------------------------------------------------------------------------
-void HCListForEach(HCListRef self, HCListForEachFunction forEachFunction) {
+void HCListForEach(HCListRef self, HCListForEachFunction forEachFunction, void* context) {
     if (forEachFunction == NULL) {
         return;
     }
     for (HCListIterator i = HCListIterationBegin(self); !HCListIterationHasEnded(&i); HCListIterationNext(&i)) {
-        forEachFunction(i.object);
+        forEachFunction(context, i.object);
     }
 }
 
-void HCListForEachWithContext(HCListRef self, HCListForEachWithContextFunction forEachFunction, void* context) {
-    if (forEachFunction == NULL) {
+void HCListFilter(HCListRef self, HCListFilterFunction filterFunction, void* context) {
+    if (filterFunction == NULL) {
+        return;
+    }
+    for (HCListIterator i = HCListIterationBeginAtLast(self); !HCListIterationHasEnded(&i); HCListIterationPrevious(&i)) {
+        if (filterFunction(context, i.object) == false) {
+            HCListRemoveObjectAtIndex(self, i.index);
+        }
+    }
+}
+
+void HCListMap(HCListRef self, HCListMapFunction mapFunction, void* context) {
+    if (mapFunction == NULL) {
         return;
     }
     for (HCListIterator i = HCListIterationBegin(self); !HCListIterationHasEnded(&i); HCListIterationNext(&i)) {
-        forEachFunction(context, i.object);
+        HCRef value = HCListRemoveObjectRetainedAtIndex(self, i.index);
+        HCRef mappedValue = mapFunction(context, value);
+        HCListAddObjectReleasedAtIndex(self, i.index, mappedValue);
+        HCRelease(value);
+    }
+}
```

Comment:
An in place transform instead of creating a new list of the transformed elements? I am not familiar with any language or library that has that behavior.

Reply:
I am fine with an in place map to transform elements because it would give you similar performance to `lazy.map` but we probably want to denote it as the in place version and have a more traditional version as well.